File: src/csharp/Foundgine.Core/Semantic/Planning/CostEstimateProvenance.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Describes where a provider cost estimate came from and how current its statistics are.</summary>
public readonly record struct CostEstimateProvenance(
    string Source,
    string? StatisticsVersion,
    DateTimeOffset? EstimatedAtUtc,
    TimeSpan? StatisticsAge,
    CostStatisticsFreshness Freshness)
{
    public static CostEstimateProvenance
        Heuristic(string source = "heuristic", DateTimeOffset? estimatedAtUtc = null) =>
        new(source, null, estimatedAtUtc ?? DateTimeOffset.UtcNow, null, CostStatisticsFreshness.Unknown);

    public static CostEstimateProvenance FromStatistics(
        string source,
        string statisticsVersion,
        DateTimeOffset observedAtUtc,
        DateTimeOffset? estimatedAtUtc = null,
        TimeSpan? staleAfter = null)
    {
        if (string.IsNullOrWhiteSpace(source)) throw new ArgumentException("Source is required.", nameof(source));
        if (string.IsNullOrWhiteSpace(statisticsVersion))
            throw new ArgumentException("Statistics version is required.", nameof(statisticsVersion));
        var now = estimatedAtUtc ?? DateTimeOffset.UtcNow;
        if (observedAtUtc > now) throw new ArgumentOutOfRangeException(nameof(observedAtUtc));
        var age = now - observedAtUtc;
        var freshness = staleAfter is null
            ? CostStatisticsFreshness.Unknown
            : age <= TimeSpan.FromTicks(staleAfter.Value.Ticks / 2)
                ? CostStatisticsFreshness.Fresh
                : age <= staleAfter.Value
                    ? CostStatisticsFreshness.Aging
                    : CostStatisticsFreshness.Stale;
        return new CostEstimateProvenance(source, statisticsVersion, now, age, freshness);
    }
}

public enum CostStatisticsFreshness
{
    Unknown = 0,
    Fresh = 1,
    Aging = 2,
    Stale = 3
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/CostEstimateProvenance.java
package com.foundgine.core.semantic.planning;

import java.time.*;

/**
 * Describes where a provider cost estimate came from and how current its
 * statistics are.
 */
public record CostEstimateProvenance(String source, String statisticsVersion, Instant estimatedAtUtc,
		Duration statisticsAge, CostStatisticsFreshness freshness) {
	public static CostEstimateProvenance heuristic() {
		return heuristic("heuristic", Instant.now());
	}

	public static CostEstimateProvenance heuristic(String source, Instant estimatedAtUtc) {
		return new CostEstimateProvenance(source, null, estimatedAtUtc, null, CostStatisticsFreshness.UNKNOWN);
	}

	public static CostEstimateProvenance fromStatistics(String source, String statisticsVersion, Instant observedAtUtc,
			Instant estimatedAtUtc, Duration staleAfter) {
		if (source == null || source.isBlank())
			throw new IllegalArgumentException("Source is required.");
		if (statisticsVersion == null || statisticsVersion.isBlank())
			throw new IllegalArgumentException("Statistics version is required.");
		Instant now = estimatedAtUtc == null ? Instant.now() : estimatedAtUtc;
		if (observedAtUtc.isAfter(now))
			throw new IllegalArgumentException("Observed time cannot be after estimate time.");
		Duration age = Duration.between(observedAtUtc, now);
		CostStatisticsFreshness f = staleAfter == null ? CostStatisticsFreshness.UNKNOWN
				: age.compareTo(staleAfter.dividedBy(2)) <= 0 ? CostStatisticsFreshness.FRESH
						: age.compareTo(staleAfter) <= 0 ? CostStatisticsFreshness.AGING
								: CostStatisticsFreshness.STALE;
		return new CostEstimateProvenance(source, statisticsVersion, now, age, f);
	}
}
